book/testing_code.md

```diff
@@ -74,152 +86,209 @@ including https://www.jeremyjordan.me/testing-ml/
 
-## Unit testing
+Likewise, tests should leave no trace. 
+Use set up and tear down code to manage this. 
+Some testing frameworks provide tools like fixtures or mocks to help with this.
 
+Tests should be independent from the code. 
+If the code changes, the tests may fail but test functions should not break. 
 
-```{admonition} Key Learning
-:class: admonition-learning
+### Tests are documentation and tests are documented
 
-You should follow the [Introduction to Unit Testing course](https://learninghub.ons.gov.uk/enrol/index.php?id=539) (GSS only) for applied examples in Python and R.
-The course also covers writing and documenting functions, and error handling.
+Good tests describe what it is they are testing. 
+They do this by using good naming practices within the test code. 
+A good test suite should tell a story - given this data, having run this function, we expect this output. 
+Tests should follow the relevant practices outlined in [core programming practices](./core_programming).
 
-Other resources include:
-* Hadley Wickham's [testthat: getting started with testing](https://vita.had.co.nz/papers/testthat.pdf)
-* [`pytest` getting started](https://docs.pytest.org/en/3.0.1/getting-started.html)
-* Real Python [Getting Started With Testing in Python](https://realpython.com/python-testing/)
-```
+The story told in a testing suite helps developers understand what code is meant to do. 
+By defining expected outputs from given inputs, developers can understand the use of a particular piece of code. 
+Good testing suites recognise this and supplement other forms of documentation.
 
-```{todo}
-These testing sections all need more content/examples.
+### Use bugs to improve your test suite
 
-[#28](https://github.com/best-practice-and-impact/qa-of-code-guidance/issues/28)
-```
+Every time you find a bug, write a test to check that it is fixed. 
+Doing so guarantees that your code covers common edge cases. 
+When you change your code or refactor your product you can be sure that bugs you have already fixed will not reappear.
 
-## Integration testing
+### Avoid flickering tests by making them deterministic
 
-Your analysis likely involves multiple units working together to perform a high level task.
-Assuring that individual units work as expected, using unit testing, does not guarantee that multiple units interact with one another as expected.
+Good tests are deterministic and repeatable. 
+Don't use randomness when writing tests. 
+The only thing that should cause a test to fail when it has been passing is if the tested code has changed. 
+Running the same test with the same tested code should always produce the same outcome.
 
-```{figure} ./_static/no_integration_tests.png
----
-width: 40%
-name: no_integration_tests
-alt: Two drawers that open into each other's handles.
----
-Two unit tests, no integration tests.
-```
+In short, tests should pass or fail, not "mostly pass".
```

Comment:
There's ambiguity in this sentence - it could be interpreted as referring to all tests, i.e. all tests should pass or all tests should fail, rather than most tests passing.

Example rewrite: "In short, unless the code changes, each test should produce the same result every time". However, I see that this isn't as punchy.